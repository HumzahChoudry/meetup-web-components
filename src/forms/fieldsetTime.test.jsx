import React from 'react';
import { shallow } from 'enzyme';
import FieldsetTime, {
	HOURS_INPUT_CLASS,
	HOURS_INPUT_NAME,
	MINUTES_INPUT_CLASS,
	MINUTES_INPUT_NAME,
	UP_ARROW,
	DOWN_ARROW,
} from './FieldsetTime';

// TODO: could use some inline snapshots here - upgrade Jest to make it happen
describe('FieldsetTime', () => {
	const MOCK_PROPS = {
		name: 'mocktime',
		value: '12:15',
		onChange: jest.fn(),
		required: true,
	};

	const render = (props = MOCK_PROPS) => shallow(<FieldsetTime {...props} />);
	[true, false].forEach(is24Hr => {
		describe(`FieldsetTime with is24hr=${is24Hr}`, () => {
			it('renders expected elements', () => {
				expect(render({ ...MOCK_PROPS, is24Hr })).toMatchSnapshot();
			});
			[
				{ name: HOURS_INPUT_NAME, className: HOURS_INPUT_CLASS },
				{ name: MINUTES_INPUT_NAME, className: MINUTES_INPUT_CLASS },
			].forEach(({ name, className }) => {
				it(`calls props.onChange with full time value when ${name} value is changed`, () => {
					MOCK_PROPS.onChange.mockClear();
					const inputTarget = {
						name,
						value: '11',
					};
					const wrapper = render({ ...MOCK_PROPS, is24Hr });
					const inputWrapper = wrapper.find(`input.${className}`);
					expect(MOCK_PROPS.onChange).not.toHaveBeenCalled();
					inputWrapper.simulate('change', {
						target: { ...inputTarget },
					});
					expect(MOCK_PROPS.onChange).toHaveBeenCalled();
					expect(MOCK_PROPS.onChange.mock.calls[0][0]).toMatchSnapshot();
				});
				[UP_ARROW, DOWN_ARROW].forEach(keyCode => {
					['00:00', '23:59'].forEach(value => {
						it(`increments/wraps ${value} on keyboard input (${keyCode}) in ${name} input`, () => {
							// this test verifies general up/down behavior as well
							MOCK_PROPS.onChange.mockClear();
							const target = {
								name,
								value: '11',
							};
							const event = {
								target,
								currentTarget: target,
								keyCode,
								preventDefault: jest.fn(),
							};
							const wrapper = render({
								...MOCK_PROPS,
								value: '00:00',
								is24Hr,
							});
							const inputWrapper = wrapper.find(`input.${className}`);
							expect(MOCK_PROPS.onChange).not.toHaveBeenCalled();
							inputWrapper.simulate('keydown', event);
							expect(MOCK_PROPS.onChange).toHaveBeenCalled();
							expect(event.preventDefault).toHaveBeenCalled();
							expect(
								MOCK_PROPS.onChange.mock.calls[0][0]
							).toMatchSnapshot();
						});
					});
				});
			});
		});
	});
	const AMPM_MAP = {
		0: 'AM',
		1: 'PM',
	};
	Object.keys(AMPM_MAP).forEach(value => {
		['12:00', '01:00', '00:00'].forEach(time => {
			it(`updates ${time} value based on meridian change to ${
				AMPM_MAP[value]
			}`, () => {
				MOCK_PROPS.onChange.mockClear();
				const target = { name: 'meridian', value };
				const wrapper = render({ ...MOCK_PROPS, is24Hr: false, value: time });
				expect(MOCK_PROPS.onChange).not.toHaveBeenCalled();
				wrapper.instance().onMeridianChange({ target });
				expect(MOCK_PROPS.onChange).toHaveBeenCalled();
				expect(MOCK_PROPS.onChange.mock.calls[0][0]).toMatchSnapshot();
			});
		});
	});
});
