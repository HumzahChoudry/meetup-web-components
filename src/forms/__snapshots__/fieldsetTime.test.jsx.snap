// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldsetTime FieldsetTime with is24hr=false calls props.onChange with full time value when hour value is changed 1`] = `
Array [
  "23:15",
]
`;

exports[`FieldsetTime FieldsetTime with is24hr=false calls props.onChange with full time value when minute value is changed 1`] = `
Array [
  "12:11",
]
`;

exports[`FieldsetTime FieldsetTime with is24hr=false renders expected elements 1`] = `
<fieldset
  className="fauxInput fauxInput--time"
>
  <Flex
    noGutters={true}
  >
    <FlexItem
      shrink={true}
    >
      <input
        className="field--reset align--center timeInput-hours"
        id="mocktime-hour"
        max={12}
        min={1}
        name="hour"
        onChange={[Function]}
        onKeyDown={[Function]}
        onMouseUp={[Function]}
        pattern="\\\\d*"
        type="number"
        value="12"
      />
       
    </FlexItem>
    <FlexItem
      className="align--center"
      shrink={true}
    >
      :
    </FlexItem>
    <FlexItem
      shrink={true}
    >
      <input
        className="field--reset align--center timeInput-minutes"
        id="mocktime-minute"
        max={59}
        min={0}
        name="minute"
        onChange={[Function]}
        onKeyDown={[Function]}
        onMouseUp={[Function]}
        pattern="\\\\d*"
        type="number"
        value="15"
      />
    </FlexItem>
    <FlexItem
      className="timeInput-meridianContainer display--flex flex--column flex--center"
      shrink={true}
    >
      <SelectInput
        className="timeInput-meridian flush--all border--none field--reset padding--left"
        id="mocktime-meridian"
        name="mwc-meridian"
        onChange={[Function]}
        options={
          Array [
            Object {
              "label": "AM",
              "value": "0",
            },
            Object {
              "label": "PM",
              "value": "1",
            },
          ]
        }
        requiredText="*"
        value="1"
      />
    </FlexItem>
  </Flex>
</fieldset>
`;

exports[`FieldsetTime FieldsetTime with is24hr=true calls props.onChange with full time value when hour value is changed 1`] = `
Array [
  "11:15",
]
`;

exports[`FieldsetTime FieldsetTime with is24hr=true calls props.onChange with full time value when minute value is changed 1`] = `
Array [
  "12:11",
]
`;

exports[`FieldsetTime FieldsetTime with is24hr=true renders expected elements 1`] = `
<fieldset
  className="fauxInput fauxInput--time"
>
  <Flex
    noGutters={true}
  >
    <FlexItem
      shrink={true}
    >
      <input
        className="field--reset align--center timeInput-hours"
        id="mocktime-hour"
        max={23}
        min={0}
        name="hour"
        onChange={[Function]}
        onKeyDown={[Function]}
        onMouseUp={[Function]}
        pattern="\\\\d*"
        type="number"
        value="12"
      />
       
    </FlexItem>
    <FlexItem
      className="align--center"
      shrink={true}
    >
      :
    </FlexItem>
    <FlexItem
      shrink={true}
    >
      <input
        className="field--reset align--center timeInput-minutes"
        id="mocktime-minute"
        max={59}
        min={0}
        name="minute"
        onChange={[Function]}
        onKeyDown={[Function]}
        onMouseUp={[Function]}
        pattern="\\\\d*"
        type="number"
        value="15"
      />
    </FlexItem>
  </Flex>
</fieldset>
`;
